src/flamenco/runtime/program/fd_bpf_loader_v3_program.c

```diff
@@ -190,1352 +330,1149 @@ fd_bpf_loader_v3_user_execute( fd_exec_instr_ctx_t ctx ) {
       }
-
-      if (FD_FEATURE_ACTIVE(ctx.slot_ctx, enable_program_redeployment_cooldown)) {
-        fd_sol_sysvar_clock_t clock;
-        fd_sysvar_clock_read( &clock, ctx.slot_ctx );
-        if (clock.slot == loader_state.inner.program_data.slot) {
-          return FD_EXECUTOR_INSTR_ERR_INVALID_ARG;
-        }
+      fd_sol_sysvar_clock_t clock = {0};
+      if( !fd_sysvar_clock_read( &clock, instr_ctx->slot_ctx ) ) {
+        FD_LOG_WARNING(( "Unable to read clock sysvar" ));
+        return FD_EXECUTOR_INSTR_ERR_GENERIC_ERR;
       }
-
-      fd_bpf_upgradeable_loader_state_t program_acc_state;
-      err = 0;
-      if (!read_bpf_upgradeable_loader_state( &ctx, program_acc, &program_acc_state, &err ))
-        return err;
-
-      if (!fd_bpf_upgradeable_loader_state_is_program( &program_acc_state )) {
+      if( clock.slot==close_account_state.inner.program_data.slot ) {
+        FD_LOG_WARNING(( "Program was deployed in this block already" ));
         return FD_EXECUTOR_INSTR_ERR_INVALID_ARG;
-      }
+      } 
 
-      if (memcmp(&program_acc_state.inner.program.programdata_address, close_acc, sizeof(fd_pubkey_t)) != 0) {
-        return FD_EXECUTOR_INSTR_ERR_INVALID_ARG;
-      }
-
-      fd_pubkey_t const * authority_acc = &txn_accs[instr_acc_idxs[2]];
-
-      err = common_close_account(ctx, authority_acc, close_acc_rec->meta, recipient_acc_rec->meta, &loader_state, close_acc);
-      if (err != 0) {
+      fd_bpf_upgradeable_loader_state_t program_state = {0};
+      err = fd_bpf_loader_v3_program_read_state( instr_ctx, program_account, &program_state );
+      if( err!=FD_BINCODE_SUCCESS ) {
+        FD_LOG_WARNING(( "Bpf state read for program account failed" ));
         return err;
       }
+      if( fd_bpf_upgradeable_loader_state_is_program( &program_state ) ) {
+        if( memcmp( &program_state.inner.program.programdata_address, close_key, sizeof(fd_pubkey_t) ) ) {
+          FD_LOG_WARNING(( "Program account does not match ProgramData account" ));
+          return FD_EXECUTOR_INSTR_ERR_INVALID_ARG;
+        }
+        err = common_close_account( close_account_state.inner.program_data.upgrade_authority_address, 
+                                    instr_ctx, 
+                                    &close_account_state );
+        if( err ) {
+          return err;
+        }
 
-      if (FD_FEATURE_ACTIVE(ctx.slot_ctx, delay_visibility_of_program_deployment)) {
-        // invoke_context.programs_modified_by_tx.replenish(
-        //     program_key,
-        //     Arc::new(LoadedProgram::new_tombstone(
-        //         clock.slot,
-        //         LoadedProgramType::Closed,
-        //     )),
-        // );
+        /* The Agave client updates the account state upon closing an account
+           in their loaded prograam cache. Checking for a program can be
+           checked by checking to see if the programdata account's loader state 
+           is unitialized. */
```

Comment:
Agave delays the visibility of program deployments: whenever a program is un-/re-/deployed, it becomes unavailable for the rest of the transaction. They do this by modifying the cache entry. Do we have these same semantics?

(See https://github.com/solana-labs/solana/pull/29654 and `delay_visibility_of_program_deployment`)

Reply:
We will add ledger tests for this